src/contracts/sgho/sGHO.sol

```diff
@@ -0,0 +1,380 @@
+    $.targetRate = 0;
+  }
+
+  /**
+   * @notice The receive function is implemented to reject direct Ether transfers to the contract.
+   * @dev sGHO does not handle ETH directly. All deposits must be made in the GHO token.
+   */
+  receive() external payable {
+    revert NoEthAllowed();
+  }
+
+  // --- ERC4626 Logic ---
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function maxWithdraw(address owner) public view override returns (uint256) {
+    if (paused()) {
+      return 0;
+    }
+
+    uint256 ghoBalance = IERC20(asset()).balanceOf(address(this));
+    uint256 maxWithdrawAssets = super.maxWithdraw(owner);
+    return maxWithdrawAssets < ghoBalance ? maxWithdrawAssets : ghoBalance;
+  }
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function maxRedeem(address owner) public view override returns (uint256) {
+    if (paused()) {
+      return 0;
+    }
+
+    uint256 ghoBalance = IERC20(asset()).balanceOf(address(this));
+    uint256 maxRedeemShares = super.maxRedeem(owner);
+    uint256 sharesForBalance = convertToShares(ghoBalance);
+    return maxRedeemShares < sharesForBalance ? maxRedeemShares : sharesForBalance;
+  }
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function maxDeposit(address) public view override returns (uint256) {
+    if (paused()) {
+      return 0;
+    }
+
+    sGHOStorage storage $ = _getsGHOStorage();
+    uint256 currentAssets = totalAssets();
+    return currentAssets >= $.supplyCap ? 0 : $.supplyCap - currentAssets;
+  }
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function maxMint(address receiver) public view override returns (uint256) {
+    return convertToShares(maxDeposit(receiver));
+  }
+
+  /**
+   * @inheritdoc IsGHO
+   */
+  function depositWithPermit(
+    uint256 assets,
+    address receiver,
+    uint256 deadline,
+    SignatureParams memory sig
+  ) external returns (uint256) {
+    // Use permit to approve the vault to spend the user's GHO tokens
+    try
+      IERC20Permit(asset()).permit(
+        _msgSender(),
+        address(this),
+        assets,
+        deadline,
+        sig.v,
+        sig.r,
+        sig.s
+      )
+    {} catch {}
+
+    return deposit(assets, receiver);
+  }
+
+  /**
+   * @inheritdoc IERC4626
+   */
+  function totalAssets() public view override returns (uint256) {
+    return _convertToAssets(totalSupply(), Math.Rounding.Floor);
+  }
+
+  /// @inheritdoc IsGHO
+  function pause() external onlyRole(PAUSE_GUARDIAN_ROLE) {
+    _pause();
+  }
+
+  /// @inheritdoc IsGHO
+  function unpause() external onlyRole(PAUSE_GUARDIAN_ROLE) {
+    _unpause();
+  }
+
+  /**
+   * @inheritdoc IsGHO
+   */
+  function setTargetRate(uint16 newRate) public onlyRole(YIELD_MANAGER_ROLE) {
+    sGHOStorage storage $ = _getsGHOStorage();
+    // Update the yield index before changing the rate to ensure proper accrual
+    if (newRate > MAX_SAFE_RATE) {
+      revert RateMustBeLessThanMaxRate();
+    }
+    _updateYieldIndex();
+    $.targetRate = newRate;
```

Comment:
why do we need to store the `targetRate` if `ratePerSecond` is the only thing that it's used for index updates?